consensus/istanbul/backend/pos.go

```diff
@@ -148,13 +146,16 @@ func (sb *Backend) updateValidatorScores(header *types.Header, state *state.Stat
 	// sb.LookbackWindow(header, state) => value at the end of epoch
 	// It doesn't matter which was the value at the beginning but how it ends.
 	// Notice that exposed metrics compute based on current block (not last of epoch) so if lookback window changed during the epoch, metric uptime score might differ
-	lookbackWindow := sb.LookbackWindow(header, state)
+	lookbackWindowFn := func() uint64 {
+		return sb.LookbackWindow(header, state)
+	}
 
-	logger = logger.New("window", lookbackWindow)
-	logger.Trace("Updating validator scores")
+	// logger = logger.New("window", lookbackWindow)
 
-	monitor := uptime.NewMonitor(store.New(sb.db), sb.EpochSize(), lookbackWindow)
-	uptimes, err := monitor.ComputeValidatorsUptime(epoch, len(valSet))
+	// monitor := uptime.NewMonitor(sb.EpochSize(), epoch, lookbackWindow, len(valSet))
+	logger.Trace("Updating validator scores")
+	monitor := sb.RetrieveUptimeScoreBuilder(header, sb.EpochSize(), lookbackWindowFn)
+	uptimes, err := monitor.ComputeUptime(header)
```

Comment:
The header and not just the bitmap? Because we validate a lot of things. Same epoch, no gaps, etc. If anything of those checks fail, the monitor wrapper, will be in charge of filling the gaps and all the other fancy things it does

Reply:
I was convinced by @hbandura that the header is helpful, as the monitor know the height of the end of the epoch, but it doesn't know the hash of it. So in the event of a reorg/setHead you need to check the hash